build.assets/robotest_run_suite.sh

```diff
@@ -3,17 +3,32 @@ set -eu -o pipefail
 
 readonly UPGRADE_FROM_DIR=${1:-$(pwd)/../upgrade_from}
 
+# decrement_patch returns x.y.(z-1) given valid x.y.z semver.
+decrement_patch() {
+    major=$(echo $1 | cut -d'.' -f1)
+    minor=$(echo $1 | cut -d'.' -f2)
+    patch=$(echo $1 | cut -d'.' -f3)
+    patch=$((patch - 1))
```

Comment:
This doesn't take pre-releases into account and also won't work for .0 patch releases (e.g. when we release 7.1.0-alpha.1, 7.1.0, etc.)

Reply:
This is a tricky one.

When I drafted this work, my plan was this bash implementation would never hit master because finding the previous testable release is challenging on master is tricky without making one or more network calls to query release metadata.

See https://github.com/gravitational/gravity/pull/1760 for my comparable master PR.

*If* this never hits master, we don't need to worry about pre-release tags or .0 tags, and can wait for a more full fledged native golang solution to #1199. However until that proper solution is available, we will have to worry about porting the change to each release branch after .0 is cut... this porting is human dependent and likely to lag behind the actual X.X.0 release -- a real risk.

Digging into this further, consider the current master state:

```bash
walt@work:~/git/gravity$ git describe --abbrev=0 master
7.1.0-alpha.1-112-g0bdbdecd
```
7.1.0-alpha.1 is tagged:

https://github.com/gravitational/gravity/releases/tag/7.1.0-alpha.1

But not released in a way that it is consumable by robotest:

```bash
walt@work:~/git/gravity$ ./e/build/current/tele pull telekube:7.1.0-alpha.1                                                                                          
Fri Jun 26 20:59:06 UTC Using Hub https://get.gravitational.io:443
Fri Jun 26 20:59:06 UTC Requesting cluster image from https://get.gravitational.io:443
Fri Jun 26 20:59:06 UTC Download finished in now
[ERROR]: package(gravitational.io/telekube:7.1.0-alpha.1) not found

```

Beyond tags without test artifacts, lets consider testing releases that aren't a traditional patch+1 bump: published X.X.X-(alpha|beta|rc), your .0 scenario, or 6.1.24 to the non-existant https://github.com/gravitational/gravity/releases/tag/6.1.23.

Looking at the previous patch release will always fail in this case, and we need to query the prior semver release.  That can't be done without knowing our release history - unlike regular x.y.z releases, the prior pre-release is impossible to predict. It could be `rc.X`, `alpha.X` or `beta.X` for any value of X.

This can be found by looking at the git tag history:

```
walt@work:~/git/gravity$ git describe master
7.1.0-alpha.1-112-g0bdbdecd
walt@work:~/git/gravity$ git describe --abbrev=0 master
7.1.0-alpha.1
```
We check to see if we're on a tag:

https://github.com/gravitational/gravity/blob/0bdbdecd9afec9938f291bf3deab1733f851ba56/version.sh#L29

And if so, looking at the git tag history instead of a naive semver patch decrement:

```bash
walt@work:~/git/gravity$ git describe --abbrev=0 7.1.0-alpha.1^
7.0.0-rc.5
```

^ This is what we'd be testing upgrade from currently on master.

This actually feels pretty reasonable and robust -- save the tagging inconsistencies in the repo.  I need to audity parity between tags & test artifacts to be confident.  I know when I checked that changelogs were consistent & sequential in https://github.com/gravitational/gravity/pull/1751 there were consistency issues:

```
missing changelogs for 6.1.23
missing changelogs for 5.5.39
missing changelogs for 5.5.30
missing changelogs for 5.5.29
missing changelogs for 5.5.25
missing changelogs for 5.5.16
missing changelogs for 5.4.8
missing changelogs for 5.0.34
missing changelogs for 5.0.31
missing changelogs for 5.0.19
missing changelogs for 5.0.5
// snip 4.x and earlier, 
````

/ramble

I'll rework this to the git history based implementation. We'll need to fix the test artifact situation for 7.1.0-alpha.1 and revisit tagging procedures in our next team meeting.

Enforcement/automation of release consistency will take work... but with the above proposal, this script will be sure to fail our PR build every time we make a release/tagging parity mistake until then.